### ISSUE TITLE
`StateCommitment` abstraction

### ISSUE THREAD

**Issue by frisitano** [paradigmxyz/reth#11830]
### Describe the feature

# Overview
This issue will be used to plan the implementation of the `StateCommitment` implementation. The `StateCommitment` will be used to provide an abstraction over the types associated with the state commitment. This will allow reth to be used with different state representations configured by the user. Initial POC implemented [here](https://github.com/paradigmxyz/reth/pull/11786).

# Plan
We will implement the `StateCommitment` via the following units of work:

- [ ] Introduce `StateCommitment` trait, integrate with `NodeTypes` and provide MPT implementation.
- [ ] Introduce `StateCommitment` types within providers.
- [ ] Introduce `HashedPostStateProvider` and refactor methods on `HashedPostState` and `PrefixSetLoader` to be generic over the `KeyHasher`

Refactor codebase to leverage `StateCommitment` types for all operations:
- [ ] StateRoot
- [ ] StorageRoot
- [ ] StateWitness
- [ ] KeyHasher

### Additional context

https://github.com/paradigmxyz/reth/pull/11786

**Comment by Thegaram** [comment]
This could be added to https://github.com/paradigmxyz/reth/issues/11241. @mattsse

**Comment by Rjected** [comment]
Just added to the tracking issue @Thegaram

**Comment by frisitano** [comment]
We have had a change of strategy at Scroll and we will no longer be using the binary Merkle Patricia trie with Poseidon key hashing. Instead, we will be migrating to native MPT, however, we would still like to land this feature in a more conservative fashion. The primary source of the invasiveness of the proposed change was the modified key hashing scheme. If we instead continue to use the standard keccak key hashing scheme the solution becomes much simpler with no need for `HashedPostStateProvider`, `HashedStorageProvider` etc. I propose that the layer at which we introduce the abstraction is the `HashBuilder` layer, this means we will introduce a trait that looks something like:

```rust
pub trait HashBuilderT {
    pub fn with_updates(mut self, retain_updates: bool) -> Self;
    pub fn with_proof_retainer(mut self, retainer: ProofRetainer) -> Self;
    pub fn set_updates(&mut self, retain_updates: bool);
    pub fn split(mut self) -> (Self, HashMap<Nibbles, BranchNodeCompact>);
    pub fn take_proof_nodes(&mut self) -> ProofNodes;
    pub fn updates_len(&self) -> usize;
    pub fn add_leaf(&mut self, key: Nibbles, value: &[u8]);
    pub fn add_branch(&mut self, key: Nibbles, value: B256, stored_in_database: bool);
    pub fn root(&mut self) -> B256;
}
```

We can then implement this trait on the different hash builder implementations as seen in the mpt and bmpt examples below:

- [mtp](https://github.com/alloy-rs/trie/blob/1ff76a618b27b66599e76c652617f398a54f0cec/src/hash_builder/mod.rs#L44-L59)
- [bmpt](https://github.com/scroll-tech/reth/blob/a337c9777863e3da3ffe9460ab3f0c2fdefc41c7/crates/scroll/trie/src/hash_builder.rs#L17-L33)

We would then make the consumers of the `HashBuilder` generic over the trait `HashBuilderT`. For example the `StateRoot`:
https://github.com/paradigmxyz/reth/blob/3e07d6575167de30739f45b014952a5d20306fce/crates/trie/trie/src/trie.rs#L21-L37

We would use the generic `HashBuilderT` type when calculating the root:

I think this yields a far less invasive change and still fulfills the function of allowing for generic state commitment. If we have an agreement on this new approach I will back out previous changes related to key hashing and then create a PR for the proposed solution.

**Comment by emhane** [comment]
thanks a lot for the update @frisitano 

@Rjected @mattsse @klkvr @rkrasiuk

**Comment by Rjected** [comment]
> HashBuilder generic over the trait HashBuilderT

I don't agree with the approach to abstract the hash builder. It's abstracting over a concrete implementation detail (the hash builder), rather than providing a good abstraction and API for implementing the state root. For example, we recently reimplemented the state root computation with what we call the sparse trie. The calculation involves:
* spawning a task
* sending it state updates
* waiting on its final result

And this results in a very different internal architecture, and a very different API, than the hash builder. If we were to introduce this abstraction, it would be difficult / impossible to implement the sparse trie. So I would prefer we go a different route, and think about what a state commitment abstraction would need to look like from first principles
